File: java/0002.两数相加.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    // 1. O(max(m, n)) t:1ms(100%) m:38.7M(32%)
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode head = new ListNode(0);
        ListNode node = head;
        int t = 0;
        while (l1 != null || l2 != null){
            int num1 = l1 != null ? l1.val : 0;
            int num2 = l2 != null ? l2.val : 0;
            t = num1 + num2 + t;
            node.next = new ListNode(t % 10);
            t /= 10;
            node = node.next;
            if (l1 != null) l1 = l1.next;
            if (l2 != null) l2 = l2.next;
        }
        if (t > 0) node.next = new ListNode(t);
        return head.next;
    }
}

File: python/0002.两数相加.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    # 1. O(max(m,n)) t:60ms(36%) m:14.9M(90%)
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> List:
        head = ListNode()
        node = head
        t = 0
        while l1 or l2:
            num1 = l1.val if l1 else 0
            num2 = l2.val if l2 else 0
            t = num1 + num2 + t
            node.next = ListNode(t % 10)
            t //= 10
            node = node.next
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
        if t == 1: # 进位情况t最多只为1
            node.next = ListNode(t)
        return head.next
